### PR TITLE
Add PostgreSQL authentication backend

### PR DESCRIPTION
**Pull request by nerzhul**
Add support for PostgreSQL authentication Database.
This is certainly the most important database backend, as it permits to share auth between servers for a server galaxy.

## To do

This PR is Ready for Review.

### CODE REVIEW

File: src/database/database-postgresql.cpp [hunk 1]
```diff
@@ -637,4 +642,174 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+		lastLoginStr.c_str()
+	};
+
+	PGresult *result = execPrepared("auth_create", 3, values, false, false);
+
+	int numrows = PQntuples(result);
+	if (numrows == 0) {
+		errorstream << "Strange behaviour on auth creation, no ID returned." << std::endl;
+		PQclear(result);
+		rollback();
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:756]: doesn't this require `beginSave` to be called first?

File: src/database/database-postgresql.cpp [hunk 2]
```diff
@@ -637,4 +642,174 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+{
+	connectToDatabase();
+}
+
+void AuthDatabasePostgreSQL::createDatabase()
+{
+	createTableIfNotExists("auth",
+		"CREATE TABLE auth ("
+			"id SERIAL,"
+			"name VARCHAR(32) UNIQUE,"
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:656]: 32 seems a bit short
In fact the [postgres documentation](https://www.postgresql.org/docs/current/datatype-character.html) suggests to just use `TEXT` in all cases.

File: src/database/database-postgresql.cpp (hunk 1 above)
Comment (nerzhul) [reply src/database/database-postgresql.cpp:756]: yep it's misplaced, my bad

File: src/database/database-postgresql.cpp (hunk 2 above)
Comment (nerzhul) [reply src/database/database-postgresql.cpp:656]: yep i prefer to use TEXT too, it was just to keep compat with sqlite3 format, but it's not a good pg practice, removing

**Comment by nerzhul** [comment]
@sfan5 it's fixed and ready for a merge

File: src/database/database-postgresql.cpp [hunk 3]
```diff
@@ -637,4 +642,174 @@ void PlayerDatabasePostgreSQL::listPlayers(std::vector<std::string> &res)
...
+			"privilege TEXT,"
+			"PRIMARY KEY (id, privilege),"
+			"CONSTRAINT fk_id FOREIGN KEY (id) REFERENCES auth (id) ON DELETE CASCADE"
+		");");
+}
+
+void AuthDatabasePostgreSQL::initStatements()
+{
+	prepareStatement("auth_read", "SELECT id, name, password, last_login FROM auth WHERE name = $1");
+	prepareStatement("auth_write", "UPDATE auth set name = $1, password = $2, last_login = $3 WHERE id = $4");
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:674]: just for consistency
```suggestion
	prepareStatement("auth_write", "UPDATE auth SET name = $1, password = $2, last_login = $3 WHERE id = $4");
```

**Pull request merged**